#### Client의 행동을 따라가보자
- Client가 어떻게 행동하는지 데이터 살펴보기

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path
import os

In [6]:
#데이터가져오기
Product = pd.read_csv("C:/Users/lsh93/Desktop/data/Product.csv")
Search1 = pd.read_csv("C:/Users/lsh93/Desktop/data/Search1.csv")
Search2 = pd.read_csv("C:/Users/lsh93/Desktop/data/Search2.csv")
Custom = pd.read_csv("C:/Users/lsh93/Desktop/data/Custom.csv")
Session = pd.read_csv("C:/Users/lsh93/Desktop/data/Session.csv")
Master = pd.read_csv("C:/Users/lsh93/Desktop/data/Master.csv")

C:\Users\lsh93\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3020: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
#구매데이터의 client 수 != 고객데이터의 client 수 
#구매데이터의 client 수 != 검색데이터의 client 수
print('Product shape: ',Product.shape)
print('Custom shape: ',Custom.shape)
print('Search1 shape: ',Search1.shape)
print('Session shape: ',Session.shape)
print()
print('구매에 참여한 client 수: ',len(Product['CLNT_ID'].value_counts()))
print('고객데이터에 있는 unique한 client 수: ',len(Custom['CLNT_ID'].value_counts()))
print('세션에 참여한 unique한 client 수: ',len(Session['CLNT_ID'].value_counts()))
print('검색에 참여한 client 수: ', len(Search1['CLNT_ID'].value_counts()))

Product shape:  (5024906, 8)
Custom shape:  (671679, 3)
Search1 shape:  (2884943, 4)
Session shape:  (2712907, 9)

구매에 참여한 client 수:  922737
고객데이터에 있는 unique한 client 수:  671679
세션에 참여한 unique한 client 수:  922737
검색에 참여한 client 수:  511477


- 세션에 참여한 client 수와 구매에 참여한 client 수는 동일하다 
- 상품을 구매했을지라도 검색을 하지 않은 client들이 있다 ) 922737 - 511477 만큼
- 고객데이터에 모든 client들의 정보가 있는 것은 아니다 

In [4]:
temp1 = pd.merge(Search1, Product, how = 'right' , on = 'CLNT_ID')

In [5]:
print(len(temp1['CLNT_ID'].value_counts()))

922737


In [6]:
temp2 = pd.merge(Custom, Product, how = 'right' , on = 'CLNT_ID')

In [7]:
print(len(temp1['CLNT_ID'].value_counts()))

922737


In [8]:
#null값 체크 --> 이건 unique한 건 아니지만 그래도 null값이 있다는 걸 의미함 
print(temp2['CLNT_AGE'].isnull().sum().sum())

1036319


- Client ID, 날짜(세션일자), 세션 ID, HIT번호, 검색키워드, 검색건수, 상품명, 구매건수 

- CLNT_ID(Product), SESS_DT(Session), SESS_ID(세션ID), HITS_SEQ(Product)
- KWD_NM(Search1), SEARCH_CNT(Search1), PD_NM(Master), PD_BUY_CT(Product)

#### Product에 있는 Client 만을 기준으로, 행동을 나열함 
#### 주의: 구매에 참여한 모든 client가 검색을 한 것은 아님

In [7]:
#구매에 참여한 client와 세션에 참여한 client의 수는 같으므로 inner join
data1 = pd.merge(Product, Session)
print(len(data1['CLNT_ID'].value_counts()))

922737


In [8]:
#상품정보를 추가해준다
data2 = pd.merge(data1, Master, how = 'left', on = 'PD_C')
print(len(data1['CLNT_ID'].value_counts()))

922737


In [11]:
#세션과 구매에 참여한 client를 기준으로 data2과 Search1을 left outer join 해 준다 
#세션과 구매에 참여한 client일지라도 검색을 하지 않은 client가 존재한다
data2 = data2[['CLNT_ID','SESS_ID','HITS_SEQ','SESS_DT','TOT_PAG_VIEW_CT','PD_NM','PD_BUY_CT','ZON_NM']]

In [14]:
idx = data2['CLNT_ID'] == 4139680
data2[idx]

CLNT_ID  SESS_ID  HITS_SEQ   SESS_DT  TOT_PAG_VIEW_CT  \
0  4139680  7605037        12  20180526             21.0   
1  4139680  7605037        12  20180526             21.0   
2  4139680  7605037        28  20180526             21.0   
3  4139680  7605037        28  20180526             21.0   
4  4139680  7605037        12  20180526             21.0   
5  4139680  7605037        28  20180526             21.0   

                        PD_NM PD_BUY_CT ZON_NM  
0               크러쉬드 립칼라 - 플럼         1  Seoul  
1  [한정판] 크러쉬드 립칼라 2.0 - 몰리 와우         1  Seoul  
2  [한정판] 크러쉬드 립칼라 2.0 - 몰리 와우         1  Seoul  
3               크러쉬드 립칼라 - 플럼         1  Seoul  
4               크러쉬드 립칼라 - 플럼         1  Seoul  
5               크러쉬드 립칼라 - 플럼         1  Seoul

In [28]:
#Session id가 양쪽 테이블에서 다르기 때문에(구매랑 검색이랑 다른 세션 아이디 가질 수도 그래서 outer로 해야 )
Client_Behavior = pd.merge(data2, Search1, how = 'outer', on = 'CLNT_ID')

In [29]:
Client_Behavior.head()

CLNT_ID  SESS_ID_x  HITS_SEQ   SESS_DT  TOT_PAG_VIEW_CT  \
0  4139680    7605037        12  20180526             21.0   
1  4139680    7605037        12  20180526             21.0   
2  4139680    7605037        28  20180526             21.0   
3  4139680    7605037        28  20180526             21.0   
4  4139680    7605037        12  20180526             21.0   

                        PD_NM PD_BUY_CT ZON_NM  SESS_ID_y KWD_NM  SEARCH_CNT  
0               크러쉬드 립칼라 - 플럼         1  Seoul        NaN    NaN         NaN  
1  [한정판] 크러쉬드 립칼라 2.0 - 몰리 와우         1  Seoul        NaN    NaN         NaN  
2  [한정판] 크러쉬드 립칼라 2.0 - 몰리 와우         1  Seoul        NaN    NaN         NaN  
3               크러쉬드 립칼라 - 플럼         1  Seoul        NaN    NaN         NaN  
4               크러쉬드 립칼라 - 플럼         1  Seoul        NaN    NaN         NaN

- client id = 4140076 일 경우

In [31]:
idx = Client_Behavior['CLNT_ID'] == 4140076
Client_Behavior[idx]

CLNT_ID  SESS_ID_x  HITS_SEQ   SESS_DT  TOT_PAG_VIEW_CT  \
6     4140076   10189797        13  20180413             18.0   
7     4140076   10189797        13  20180413             18.0   
8     4140076   10189797        13  20180413             18.0   
9     4140076   10189797        13  20180413             18.0   
10    4140076   10189797        13  20180413             18.0   
11    4140076   10189797        13  20180413             18.0   
12    4140076   10189797        13  20180413             18.0   
13    4140076   10189797        13  20180413             18.0   
14    4140076   10189797        13  20180413             18.0   
15    4140076   10189797        13  20180413             18.0   
16    4140076   10189797        13  20180413             18.0   
17    4140076   10189797        13  20180413             18.0   
18    4140076   10189797        13  20180413             18.0   
19    4140076   10189797        13  20180413             18.0   
20    4140076   10189797        13  20180413             18.0   
21    4140076   10189797        13  20180413             18.0   
22    4140076   10189797        13  20180413             18.0   
23    4140076   10189797        13  20180413             18.0   
24    4140076   10189797        13  20180413             18.0   
25    4140076   10189797        13  20180413             18.0   
26    4140076   10189797        13  20180413             18.0   
27    4140076   10189797        13  20180413             18.0   
28    4140076   10189797        13  20180413             18.0   
29    4140076   10189797        13  20180413             18.0   
30    4140076   10189797        13  20180413             18.0   
31    4140076   10189797        13  20180413             18.0   
32    4140076   10189797        13  20180413             18.0   
33    4140076   10189797        13  20180413             18.0   
34    4140076   10189797        13  20180413             18.0   
35    4140076   10189797        13  20180413             18.0   
...       ...        ...       ...       ...              ...   
2680  4140076    7803570        51  20180523             83.0   
2681  4140076    7803570        51  20180523             83.0   
2682  4140076    7803570        51  20180523             83.0   
2683  4140076    7803570        51  20180523             83.0   
2684  4140076    7803570        51  20180523             83.0   
2685  4140076    7803570        51  20180523             83.0   
2686  4140076    7803570        51  20180523             83.0   
2687  4140076    7803570        51  20180523             83.0   
2688  4140076    7803570        51  20180523             83.0   
2689  4140076    7803570        51  20180523             83.0   
2690  4140076    7803570        51  20180523             83.0   
2691  4140076    7803570        51  20180523             83.0   
2692  4140076    7803570        51  20180523             83.0   
2693  4140076    7803570        51  20180523             83.0   
2694  4140076    7803570        51  20180523             83.0   
2695  4140076    7803570        51  20180523             83.0   
2696  4140076    7803570        51  20180523             83.0   
2697  4140076    7803570        51  20180523             83.0   
2698  4140076    7803570        51  20180523             83.0   
2699  4140076    7803570        51  20180523             83.0   
2700  4140076    7803570        51  20180523             83.0   
2701  4140076    7803570        51  20180523             83.0   
2702  4140076    7803570        51  20180523             83.0   
2703  4140076    7803570        51  20180523             83.0   
2704  4140076    7803570        51  20180523             83.0   
2705  4140076    7803570        51  20180523             83.0   
2706  4140076    7803570        51  20180523             83.0   
2707  4140076    7803570        51  20180523             83.0   
2708  4140076    7803570        51  20180523             83.0   
2709  4140076    7803570        51  20180

In [32]:
#Session이랑 Search1, Product에서 같은 client id에 대해서 SESS_ID비교,,, 뭔가 이상해서 
idx = Session['CLNT_ID'] == 4140076
Session[idx]

CLNT_ID   SESS_ID  SESS_SEQ   SESS_DT  TOT_PAG_VIEW_CT TOT_SESS_HR_V  \
880512   4140076  10767546       139  20180404             19.0           467   
880513   4140076   8891563       224  20180505             19.0           172   
1464484  4140076   6517872       273  20180613            179.0         1,563   
1464485  4140076   9147546       213  20180501            473.0         8,164   
1464486  4140076   2051362       304  20180827             78.0         1,992   
1464487  4140076   8326529       229  20180514            199.0         2,055   
1464488  4140076   7782189       241  20180523            128.0         2,066   
1464489  4140076  10760242       140  20180404             26.0            54   
1464490  4140076  10574005       146  20180407            257.0         1,201   
1464491  4140076  10584304       145  20180407             79.0         1,400   
1464492  4140076   9992238       183  20180417            112.0         1,110   
1464493  4140076   6591360       264  20180612            149.0           843   
1464494  4140076   7754761       242  20180524             71.0         1,320   
1464495  4140076   8608110       226  20180510            366.0         2,903   
1464496  4140076   5384543       285  20180702             45.0           191   
1464497  4140076   8235168       230  20180516            135.0           733   
1464498  4140076   8153697       231  20180517            132.0         1,152   
1464499  4140076   6885554       253  20180607            133.0           880   
1464500  4140076   9899980       187  20180418             34.0           809   
1464501  4140076   6526651       272  20180613            192.0         2,147   
1464502  4140076  10883079       132  20180402             29.0            82   
1464503  4140076   6538217       269  20180613             26.0           205   
1464504  4140076   4756252       298  20180712            146.0         1,275   
2666011  4140076   5338950       287  20180703             81.0           491   
2666012  4140076  10342512       156  20180411             37.0           873   
2704747  4140076   7803570       239  20180523             83.0           467   
2704748  4140076  10189797       172  20180413             18.0           113   

        DVC_CTG_NM       ZON_NM     CITY_NM  
880512      mobile        Seoul       Seoul  
880513      mobile  Gyeonggi-do   Icheon-si  
1464484     mobile  Gyeonggi-do   Icheon-si  
1464485     mobile  Gyeonggi-do   Icheon-si  
1464486     mobile        Seoul       Seoul  
1464487     mobile  Gyeonggi-do   Icheon-si  
1464488     mobile  Gyeonggi-do   Icheon-si  
1464489     mobile  Gyeonggi-do   Icheon-si  
1464490     mobile        Seoul       Seoul  
1464491     mobile  Gyeonggi-do   Icheon-si  
1464492     mobile        Seoul       Seoul  
1464493     mobile        Seoul       Seoul  
1464494     mobile  Gyeonggi-do   Icheon-si  
1464495     mobile        Seoul       Seoul  
1464496     mobile  Gyeonggi-do   Icheon-si  
1464497     mobile  Gyeonggi-do  Gwangju-si  
1464498     mobile  Gyeonggi-do  Gwangju-si  
1464499     mobile        Seoul       Seoul  
1464500     mobile  Gyeonggi-do  Gwangju-si  
1464501     mobile        Seoul       Seoul  
1464502     mobile  Gyeonggi-do   Icheon-si  
1464503     mobile  Gyeonggi-do   Icheon-si  
1464504     mobile  Gyeonggi-do   Icheon-si  
2666011     mobile  Gyeonggi-do   Icheon-si  
2666012     mobile  Gyeonggi-do  Gwangju-si  
2704747     mobile        Seoul       Seoul  
2704748     mobile  Gyeonggi-do   Icheon-si

In [33]:
#Session이랑 Search1, Product에서 같은 client id에 대해서 SESS_ID비교,,, 뭔가 이상해서 
idx = Search1['CLNT_ID'] == 4140076
Search1[idx]

CLNT_ID   SESS_ID          KWD_NM  SEARCH_CNT
365805   4140076  10342512             크록스           1
365806   4140076   5338950            여성백팩           2
434665   4140076   7803570            닥스여아           1
434666   4140076   7803570             프라다           1
1861286  4140076   4756252             데코뷰           1
1861287  4140076   6526651          나이키슬리퍼           1
1861288  4140076   6526651        프레드페리 남성           2
1861289  4140076   6526651           게스 남성           1
1861290  4140076   8326529         아디다스반바지           1
1861291  4140076   4756252               뷰           1
1861292  4140076   9147546           나이키키즈           1
1861293  4140076   9992238          스톤아일랜드           1
1861294  4140076  10883079       슈퍼콤마비 운동화           1
1861295  4140076   6526651            빈폴남성           1
1861296  4140076   9147546            빈폴키즈           1
1861297  4140076  10574005     아쿠아슈즈 노스페이스           3
1861298  4140076   4756252            베베데코           1
1861299  4140076   9992238       꼼데가르송 가디건           1
1861300  4140076   8235168         아디다스반바지           1
1861301  4140076   2051362           프레드페리           1
1861302  4140076  10574005      아쿠아슈즈 아디다스           1
1861303  4140076  10574005       아쿠아슈즈 나이키           1
1861304  4140076   6591360        프레드페리 남성           1
1861305  4140076   6591360            게스남성           1
1861306  4140076   9147546             콩순이           1
1861307  4140076   5384543        데코뷰 암막커튼           1
1861308  4140076   9147546             콩콩이           1
1861309  4140076   9147546          뉴발란스키즈           1
1861310  4140076   9147546        닌텐도스위치게임           1
1861311  4140076   6885554        프레드페리 남성           1
1861312  4140076   8153697        로즈몽 여성시계           1
1861313  4140076   5384543          데코뷰 커튼           1
1861314  4140076   6885554    프레드페리 남성 티셔츠           1
1861315  4140076   8153697  파리게이츠골프 티셔츠 남성           2
1861316  4140076  10584304           여성블로퍼           1
1861317  4140076   9147546       닌텐도스위치 게임           2
1861318  4140076  10574005            남아팬티           1
1861319  4140076   4756252           빨래바구니           2
1861320  4140076   9147546         어린이날 선물           1
1861321  4140076   6517872             핏플랍           1
1861322  4140076   8153697           파리게이츠           2
1861323  4140076  10760242       슈퍼콤마비 운동화           1
1861324  4140076   6591360          크록스 남성           1
1861325  4140076   6517872           슈콤마보니           1
1861326  4140076  10584304           슈퍼콤마비           2
1861327  4140076   9147546          아디다스키즈           2
1861328  4140076   6591360        파리게이츠 남성           1
1861329  4140076   7782189         톰브라운 남성           1
1861330  4140076  10584304           주니어팬티           1
1861331  4140076   9147546        프레드페리티셔츠           5
1861332  4140076   9992238           지방시가방           1
1861333  4140076   9992238              구찌           1

In [34]:
idx = Product['CLNT_ID'] == 4140076
Product[idx]

CLNT_ID   SESS_ID  HITS_SEQ    PD_C  \
1        4140076  10189797        13  570603   
132425   4140076  10342512        23  422358   
647279   4140076   6526651        22  196272   
673008   4140076   8326529       199  305965   
720869   4140076   8891563        19  305965   
783221   4140076   6526651       193  196272   
784254   4140076   6591360       149  196272   
844514   4140076   6526651        22  365025   
1062425  4140076   4756252       147  794908   
1168519  4140076  10342512        23  403243   
1336145  4140076   9147546       474  544957   
1667666  4140076   6526651        22  364992   
1892440  4140076   6591360       149  628713   
2002362  4140076  10574005       253  450131   
2245769  4140076   9147546       462  533783   
2290388  4140076   6526651        22  339321   
2417403  4140076  10767546        19  600758   
2474452  4140076  10767546        19  471477   
2521278  4140076   7754761        71  315988   
2534761  4140076   6517872       179  373461   
2604827  4140076   5338950        76  598881   
2618571  4140076  10883079        29  689663   
2626331  4140076  10584304        50  689663   
2764356  4140076  10760242        26  689663   
2792404  4140076   6885554       134  288192   
2919327  4140076   6885554       118  522552   
2973072  4140076   9992238       112  495110   
3078609  4140076   8153697       123  555694   
3106144  4140076   6591360       149  526369   
3251603  4140076  10342512        23  806382   
3269915  4140076   4756252        76  570089   
3276182  4140076   2051362        76  195687   
3298213  4140076   6538217        26  739858   
3361106  4140076   9899980        34  606095   
3409826  4140076   7782189       125  606096   
3457190  4140076  10574005       253  481952   
3924326  4140076   6526651       193  507637   
3939434  4140076   6591360       149  507637   
4061040  4140076   8608110       366  447887   
4083909  4140076   8235168       135  447887   
4083910  4140076   8891563        19  447887   
4098729  4140076   8326529       199  447887   
4282873  4140076   5384543        45  489476   
4282874  4140076  10189797        13  411029   
4310534  4140076   6526651       193  262956   
4489275  4140076   8608110       366  567645   
4496784  4140076   8235168       135  275704   
4526058  4140076   8235168       135  567645   
4537873  4140076   8326529       199  275704   
4544962  4140076  10574005       253  670971   
4869790  4140076   8608110       366  447893   
4944569  4140076   7803570        51  453110   

                                   PD_ADD_NM     PD_BRA_NM PD_BUY_AM PD_BUY_CT  
1           색상:BLK0_(BLK0)BLACK|사이즈:120 / 2개           데상트    39,000         2  
132425   선택:02.204143-02S|사이즈:M5W7(240) / 1개           크록스    39,900         1  
647279       색상:선택3)248(네이비)|사이즈:L(105) / 1개         프레드페리    47,600         1  
673008                      사이즈:115_3XL / 1개      아디다스(의류)    46,700         1  
720869                      사이즈:115_3XL / 1개      아디다스(의류)    46,700         1  
783221       색상:선택2)102(블랙)|사이즈:XL(110) / 1개         프레드페리    47,600         1  
784254       색상:선택2)102(블랙)|사이즈:XL(110) / 2개         프레드페리    47,600         2  
844514                    색상:WH|사이즈:165 / 1개         닥스 키즈    48,300         1  
1062425                                   1개         네이쳐리빙     7,200         1  
1168519                          사이즈:M9 / 1개           크록스    57,000         1  
1336145                                   1개           영실업    59,820         1  
1667666                   색상:BG|사이즈:165 / 1개         닥스 키즈    71,200         1  
1892440               사이즈:M7(260mm)_620 / 1개           크록스    69,900         1  
2002362                  색상:BLK|사이즈:080 / 1개      아레나(수영복)    79,200         1  
2245769                     사이즈:XL(110) / 1개         프레드페리    85,000         1  
2290388                   색상:BU|사이즈:165 / 1개         닥스 키즈    90,300         1  
2417403                      사이즈:L(105) / 1개         프레드페리   103,600         1  
247445